TombEngine/Objects/Effects/Fireflies.cpp

```diff
@@ -0,0 +1,356 @@
+#include "framework.h"
+#include "Objects/Effects/Fireflies.h"
+
+#include "Game/collision/collide_item.h"
+#include "Game/collision/collide_room.h"
+#include "Game/collision/Point.h"
+#include "Game/control/box.h"
+#include "Game/control/flipeffect.h"
+#include "Game/effects/effects.h"
+#include "Game/effects/Streamer.h"
+#include "Game/effects/tomb4fx.h"
+#include "Game/items.h"
+#include "Game/Lara/lara.h"
+#include "Game/Lara/lara_helpers.h"
+#include "Game/misc.h"
+#include "Game/Setup.h"
+#include "Math/Math.h"
+#include "Renderer/Renderer.h"
+#include "Specific/clock.h"
+#include "Specific/level.h"
+
+using namespace TEN::Collision::Point;
+using namespace TEN::Math;
+using namespace TEN::Renderer;
+using namespace TEN::Effects::Streamer;
+
+namespace TEN::Effects::Fireflys
+{
+    constexpr auto FIREFLY_COHESION_FACTOR = 600.1f;
+    constexpr auto FIREFLY_SPACING_FACTOR = 600.0f;
+    constexpr auto FIREFLY_CATCH_UP_FACTOR = 0.2f;
+    constexpr auto FIREFLY_TARGET_DISTANCE_MAX = SQUARE(BLOCK(1.0f));
+    constexpr auto FIREFLY_BASE_SEPARATION_DISTANCE = 10.0f;
+    constexpr auto FIREFLY_FLEE_DISTANCE = BLOCK(0.7);
+    constexpr auto FIREFLY_RETURN_DISTANCE = BLOCK(4);
+    constexpr auto MAX_FIREFLIES = 92;
+    constexpr auto DEFAULT_FIREFLY_COUNT = 24;
+    constexpr auto FIREFLY_RISE_UP_FACTOR = 200;
+
+    std::vector<FireflyData> FireflySwarm = {};
+    std::unordered_map<int, int> nextFireflyNumberMap; // Numbering the Fireflies for Streamer effect.
+
+    void InitializeFireflySwarm(short itemNumber)
+    {
+        auto& item = g_Level.Items[itemNumber];
+
+        item.Animation.Velocity.z = Random::GenerateFloat(32.0f, 160.0f);
+
+        item.HitPoints = DEFAULT_FIREFLY_COUNT;
+        item.ItemFlags[FirefliesItemFlags::TargetItemPtr] = item.Index;
+        item.ItemFlags[FirefliesItemFlags::Light] = 1; // 0 = Turn off light effect, 1 = turn on light (DEFAULT).
+
+        item.ItemFlags[FirefliesItemFlags::TriggerFlags] = item.TriggerFlags;
```

Comment:
You can replace this and two if conditions down the line with `item.ItemFlags[FirefliesItemFlags::TriggerFlags] = std::clamp(item.TriggerFlags, -8, 8);`

Reply:
fixed